src/net/sourceforge/kolmafia/request/RelayRequest.java

```diff
@@ -3500,6 +3515,191 @@ public boolean pollForCompletion(String type) {
+   * (objectType, identifierString, identifierNumber if applicable) set.
+   *
+   * <p>Details for an enumerated type instance can be requested with the special "identity"
+   * function, which takes one placeholder argument.
+   *
+   * @param request Request body
+   */
+  private void handleJsonApi(String request) {
+    this.contentType = "application/json";
+    JSONObject json;
+    try {
+      json = JSON.parseObject(request);
+    } catch (JSONException e) {
+      jsonError("Invalid JSON object in request.");
+      return;
+    }
+
+    var result = new JSONObject();
+
+    var properties = json.get("properties");
+    if (properties instanceof JSONArray propertiesArray) {
+      var nonString = propertiesArray.stream().filter(name -> !(name instanceof String)).toList();
+      if (!nonString.isEmpty()) {
+        jsonError("Invalid property names " + JSON.toJSONString(nonString));
+        return;
+      }
+      // Now all elements must be strings.
+      result.put(
+          "properties",
+          propertiesArray.stream().map(name -> Preferences.getString((String) name)).toList());
+    }
+
+    var functions = json.get("functions");
+    if (functions instanceof JSONArray functionsArray) {
+      var nonMatching =
+          functionsArray.stream()
+              .filter(
+                  func -> {
+                    if (!(func instanceof JSONObject funcObject)) return true;
+                    var name = funcObject.get("name");
+                    var args = funcObject.get("args");
+                    if (!(name instanceof String) || !(args instanceof JSONArray)) return true;
+                    return ((JSONArray) args).contains(null);
+                  })
+              .toList();
+      if (!nonMatching.isEmpty()) {
+        jsonError("Invalid function calls " + JSON.toJSONString(nonMatching));
+        return;
+      }
+
+      // Everything validated. Transform all function arguments.
+      var functionsResult = new JSONArray(functionsArray.size());
+      for (var func : functionsArray) {
+        var funcObject = (JSONObject) func;
+        var name = (String) funcObject.get("name");
+        var args = (JSONArray) funcObject.get("args");
+        var transformedArguments = args.stream().map(this::transformApiArgument).toList();
+        var badIndices =
+            IntStream.range(0, transformedArguments.size())
+                .filter(i -> transformedArguments.get(i) == null)
+                .toArray();
+        if (badIndices.length > 0) {
+          jsonError(
+              "Invalid arguments to "
+                  + name
+                  + ": "
+                  + JSON.toJSONString(Arrays.stream(badIndices).mapToObj(args::get)));
+          return;
+        }
+
+        if (name.equals("identity") && args.size() == 1) {
+          functionsResult.add(
+              transformResult(JSONValueConverter.asJSON(transformedArguments.get(0).asProxy())));
```

Comment:
what happens if you call "identity" on something that isn't an enumerated type (e.g. you call it on a string)?

What should happen?